### PR TITLE
[jit] Fix alias analysis for fork/wait

### PR DESCRIPTION
**Pull request by suo**
(review top commit only).

As expected, fork/wait introduces some corner cases into the alias analysis. The comments inline should describe the changes.

### CODE REVIEW

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 1]
```diff
@@ -496,6 +513,66 @@ void AliasDb::analyzeChunk(Node* node) {
...
+    // Say we define the "live interval" of a fork the interval between the
+    // `fork` and its first corresponding `wait` (inclusive).
+    //
+    // Any writes in the subgraph can happen at any point in the live interval,
+    // so it's not safe to re-order any reads to those memory locations from
+    // outside the live interval to inside.
+    //
+    // In reality, any reads *inside* the live interval are undefined behavior,
+    // since the writes may or may not have been executed yet. But we'll let
+    // users do that and shoot themselves in the foot for now.
```
Comment (jamesr66a) [inline torch/csrc/jit/passes/alias_analysis.cpp]: Do we want to print a warning in this situation?

**Review by jamesr66a — COMMENTED** [review]
LGTM, just had one question inside

File: torch/csrc/jit/passes/alias_analysis.cpp (hunk 1 above)
Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.cpp]: I'm reluctant to incorporate alias analysis results in user-visible error reporting. We can't expect users to know about alias info/the IR/how we've mutated the graph up till now, so an error msg would be super confusing and probably not actionable.

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 2]
```diff
@@ -496,6 +513,66 @@ void AliasDb::analyzeChunk(Node* node) {
   }
 }
 
+void AliasDb::analyzeWait(Node* node) {
+  // Do nothing, this node should have already been analyzed by the
+  // corresponding analyzeFork(). See that function for details.
+}
+
+void AliasDb::analyzeFork(Node* node) {
+  // We treat forks as if their subgraph ran immediately. Since the alias
```
Comment (eellison) [inline torch/csrc/jit/passes/alias_analysis.cpp]: Maybe rephrase ? We're treating forks as potentially running any time being fork & wait

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 3]
```diff
@@ -484,6 +513,59 @@ void AliasDb::analyzeChunk(Node* node) {
...
+  analyze(subgraphBlock);
+
+  // Propagate aliasing and write information from the subgraph outputs to the
+  // outputs of the corresponding aten::wait() calls, since that's where the
+  // values will eventually emerge.
+  const auto fut = node->output();
+  const auto subgraphWrites = getWrites(subgraph->block());
+  for (const auto& use : fut->uses()) {
+    const auto wait = use.user;
+    AT_ASSERT(wait->outputs().size() == subgraph->outputs().size());
```
Comment (eellison) [inline torch/csrc/jit/passes/alias_analysis.cpp]: Will this fail is it's used in an if node ?

Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.cpp]: Hmm…good point. I guess it's non-trivial to find all the matching "waits".

**Review by jamesr66a — APPROVED** [review]
ok google

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 4]
```diff
@@ -20,6 +21,42 @@ bool shouldAnnotate(const TypePtr& type) {
 bool shouldAnnotate(const Value* v) {
   return shouldAnnotate(v->type());
 }
+
+std::unordered_set<Node*> getMatchingWaits(Value* fut) {
```
Comment (eellison) [inline torch/csrc/jit/passes/alias_analysis.cpp]: haven't taken full look at new changes but It doesn't look like this is being called except recursively

Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.cpp]: blah, this was left in, will remove!
